internal/bridge/managedpostgrescluster/managedpostgrescluster_controller.go

```diff
@@ -0,0 +1,552 @@
+				Message: fmt.Sprintf(
+					"Cannot create from spec for some reason: %v", "TODO(managedpostgrescluster)"),
+			})
+
+			// TODO(managedpostgrescluster): If the payload is wrong, we don't want to requeue, so pass nil error
+			// If the transmission hit a transient problem, we do want to requeue
+			return ctrl.Result{}, nil
+		}
+		managedpostgrescluster.Status.ID = cluster.ID
+		return ctrl.Result{RequeueAfter: 3 * time.Minute}, nil
+	}
+
+	// If we reach this point, our ManagedPostgresCluster object has an ID
+	// so we want to fill in the details for the cluster and cluster upgrades from the Bridge API
+	// Consider cluster details as a separate func.
+	clusterDetails, err := r.NewClient().GetCluster(ctx, key, managedpostgrescluster.Status.ID)
+	if err != nil {
+		log.Error(err, "whoops, cluster getting issue")
+		return ctrl.Result{}, err
+	}
+	managedpostgrescluster.Status.Cluster = clusterDetails
+
+	clusterUpgradeDetails, err := r.NewClient().GetClusterUpgrade(ctx, key, managedpostgrescluster.Status.ID)
+	if err != nil {
+		log.Error(err, "whoops, cluster upgrade getting issue")
+		return ctrl.Result{}, err
+	}
+	managedpostgrescluster.Status.ClusterUpgrade = clusterUpgradeDetails
+
+	// Requeue if state creating?
+
+	// For now just skip updating until the upgrade is cleared
+	// For future, how to tell that an upgrade in progress is the one we want?
+	// Perhaps add `generation` field to upgrade
+	if len(managedpostgrescluster.Status.ClusterUpgrade.Operations) != 0 {
+		return ctrl.Result{RequeueAfter: 3 * time.Minute}, nil
+	}
+
+	// Consider upgrade checking as a func
+	// Consider upgrading as a func
+
+	// Are there diffs between the cluster response from the Bridge API and the spec?
+	// HA diffs are sent to /clusters/{cluster_id}/actions/[enable|disable]-ha
+	// so have to know (a) to send and (b) which to send to
+	haDiff := managedpostgrescluster.Spec.IsHA != managedpostgrescluster.Status.Cluster.IsHA
+
+	// Check if there's an upgrade difference for the three upgradeable fields that hit the upgrade endpoint
+	val, _ := managedpostgrescluster.Spec.Storage.AsInt64()
+	upgradeDiff := (val != int64(managedpostgrescluster.Status.Cluster.Storage) ||
+		managedpostgrescluster.Spec.Plan != managedpostgrescluster.Status.Cluster.Plan ||
+		managedpostgrescluster.Spec.PostgresVersion != managedpostgrescluster.Status.Cluster.MajorVersion)
+
+	// Check if there's a difference in is_protected, name, maintenance_window_start
+	// the three fields that hit the PATCH `clusters/<id>` endpoint
+	// TODO(managedpostgrescluster)
+
+	if upgradeDiff {
+		return r.handleUpgrade(ctx, key, managedpostgrescluster, val)
+	}
+	if haDiff {
+		return r.handleUpgradeHA(ctx, key, managedpostgrescluster)
+	}
+
+	// Doy, use the result to get an accurate log
+	log.Info("Reconciled", "requeue", err != nil ||
+		result.Requeue ||
+		result.RequeueAfter > 0)
+	// Always requeue?
+	return ctrl.Result{RequeueAfter: 3 * time.Minute}, nil
```

Comment:
Maybe doing this means we don't need the Watch?

Reply:
Maybe add a TODO questioning whether we want this or the Watch or both?